### ISSUE TITLE
[BOOM-3633] ps_mainmenu showing items where items excluded with group access

### ISSUE THREAD

**Issue by prestonBot** [PrestaShop/PrestaShop#9924]
> This issue has been migrated from the Forge. Read the [original ticket here](http://forge.prestashop.com/browse/BOOM-3633).

- _**Reporter:**_ dubaku
- _**Created at:**_ Thu, 10 Aug 2017 09:56:02 +0200

<p>Fresh install of 1.7, I made 2 category where the first and the second have different group access set (the first category visitors can see, the second not), but in the main menu every two cat is showing, just when I want to go inside the second category is a message "You do not have access to this category"</p>

<p>I need to set to some category be showed and others are not. Why is this not working, what I missed?</p>
        
* How to reproduce the issue ?

Set a few category with different group access

**Comment by prestonBot** [comment]
> This comment has been migrated from the Forge. Read the [original comment here](http://forge.prestashop.com/browse/BOOM-3633#comment-151733).

- _**Author:**_ dubaku
- _**Created at:**_ Thu, 10 Aug 2017 09:57:29 +0200

<p>in 1.6 the module name was blocktopmenu and there wasn't any problem with this</p>

**Comment by prestonBot** [comment]
> This comment has been migrated from the Forge. Read the [original comment here](http://forge.prestashop.com/browse/BOOM-3633#comment-151801).

- _**Author:**_ prasanth
- _**Created at:**_ Fri, 11 Aug 2017 11:45:19 +0200

<p>Hi ikedubaku,</p>

<p>In your example, this message is shown only when a customer is logged and is on his access page then log out. If I'm a visitor the category is not shown.<br/>
See group_category.gif<br/>
Is it the same for you?</p>

**Comment by prestonBot** [comment]
> This comment has been migrated from the Forge. Read the [original comment here](http://forge.prestashop.com/browse/BOOM-3633#comment-151807).

- _**Author:**_ dubaku
- _**Created at:**_ Fri, 11 Aug 2017 13:39:08 +0200

<p>Hi Prasanth Selvarasa,</p>

<p>in your example the side menu is in focus, I believe that's another module. (Category tree links:ps_categorytree)<br/>
But there is an interesting thing, this module is doing the filtering(group access), what ps_mainmenu don't.<br/>
It's should be easy to crosscheck this two module for someone who have more programming skills than me, what is missing from the ps_mainmenu.<br/>
Thanks</p>

**Comment by prestonBot** [comment]
> This comment has been migrated from the Forge. Read the [original comment here](http://forge.prestashop.com/browse/BOOM-3633#comment-152053).

- _**Author:**_ prasanth
- _**Created at:**_ Thu, 17 Aug 2017 16:59:35 +0200

<p>En 1.7, deux modules peuvent gérer l'affichage des catégories en Front : <b>Main menu</b> et <b>Category tree links</b>. Lorsque qu'on limite l'accès d'une catégorie à un certain groupe on a un comportement différent en FO selon les deux modules : </p>
<ul class="alternate" type="square">
	<li><b>Category tree links</b> : la catégorie est visible en FO seulement pour les groupes ayant accès</li>
	<li><b>Main menu</b> : la categorie est visible pour tout le monde avec un message d'erreur si on est pas censé voir le contenu</li>
</ul>


<p>Le comportement voulu est que la catégorie ne soit pas affiché en FO c'est à dire l'actuel comportement de <b>Category tree links</b>. Il faudrai donc l'adapter également au module <b>Main menu</b></p>

<p>Par ailleurs, le message d'erreur n'est pas correctement affiché, il devrait plutôt s'afficher dans l'espace où les produits s'affichent. (access_category_error.png)</p>

**Comment by prestonBot** [comment]
> This comment has been migrated from the Forge. Read the [original comment here](http://forge.prestashop.com/browse/BOOM-3633#comment-160835).

- _**Author:**_ dubaku
- _**Created at:**_ Tue, 23 Jan 2018 13:37:17 +0100

<p>Hi,</p>

<p>how is possible that since 5 month nothing changed on this case?</p>

<p>Regards</p>

**Comment by prestonBot** [comment]
> This comment has been migrated from the Forge. Read the [original comment here](http://forge.prestashop.com/browse/BOOM-3633#comment-162505).

- _**Author:**_ dubaku
- _**Created at:**_ Tue, 20 Feb 2018 11:24:11 +0100

<p>I see there is other things to do, but hey, the navigation depending on group access policy is a base-main function, it's urgent to fix.</p>

<p><b>I can't use a car without a steering wheel!</b></p>

<p>The navigation settings are in the software by default and these are there for cause. 1.7 is not an upgrade, when it's worse than 1.6.</p>

<p>I'm still waiting for this small fix.</p>

**Comment by prestonBot** [comment]
> This comment has been migrated from the Forge. Read the [original comment here](http://forge.prestashop.com/browse/BOOM-3633#comment-171544).

- _**Author:**_ ikaemos
- _**Created at:**_ Fri, 13 Jul 2018 01:45:22 +0200

<p>Hi,</p>

<p>To prevent "ps_mainmenu" from displaying to wrong category, I applied this (need to test with banned customer from geolocation) :</p>

<p>In "Ps_MainMenu" class:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
<span class="code-keyword">protected</span> function getCacheDirectory()
{
    $groups = (isset($<span class="code-keyword">this</span>->context->customer)) ? $<span class="code-keyword">this</span>->context->customer->getGroups() : <span class="code-keyword">null</span>;
    $dir =_PS_CACHE_DIR_ . <span class="code-quote">'ps_mainmenu'</span> . (((isset($groups)) && (is_array($groups)) && (count($groups) > 0)) ? <span class="code-quote">'/'</span> . implode(<span class="code-quote">'_'</span>, $groups) : '');
    <span class="code-keyword">if</span> (!(is_dir($dir)))
mkdir($dir, 0775, <span class="code-keyword">true</span>);

    <span class="code-keyword">return</span> $dir;
}

<span class="code-keyword">protected</span> function clearMenuCache()
{
    $dir = $<span class="code-keyword">this</span>->getCacheDirectory();
    <span class="code-keyword">if</span> (!(is_dir($dir)))
<span class="code-keyword">return</span>;

    $<span class="code-keyword">this</span>->cleanMenuCacheDirectory($dir);
}

<span class="code-keyword">private</span> function cleanMenuCacheDirectory(string $dir)
{
    foreach (scandir($dir) as $entry)
    {
<span class="code-keyword">if</span> (($entry !== <span class="code-quote">'.'</span>) && ($entry !== <span class="code-quote">'..'</span>))
{
    $path = $dir . DIRECTORY_SEPARATOR . $entry;
    <span class="code-keyword">if</span> (is_dir($path))
$<span class="code-keyword">this</span>->cleanMenuCacheDirectory($path);
    <span class="code-keyword">else</span> <span class="code-keyword">if</span> (preg_match(<span class="code-quote">'/\.json$/'</span>, $entry))
unlink($path);
}
    }
}
</pre>
</div></div>
<p>Then is "generateCategoriesMenu" method:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
foreach ($categories as $key => $category)
{
    $node = $<span class="code-keyword">this</span>->makeNode([]);
    <span class="code-keyword">if</span> ($category[<span class="code-quote">'level_depth'</span>] > 1)
    {
$cat = <span class="code-keyword">new</span> Category($category[<span class="code-quote">'id_category'</span>]);
$link = $cat->getLink();
<span class="code-comment">// Check <span class="code-keyword">if</span> customer is set and check access
</span><span class="code-keyword">if</span> ((isset($<span class="code-keyword">this</span>->context->customer)) && (!($cat->checkAccess($<span class="code-keyword">this</span>->context->customer->id)))) {
    <span class="code-keyword">continue</span>;
}
    }
    <span class="code-keyword">else</span>
    {
$link = $<span class="code-keyword">this</span>->context->link->getPageLink(<span class="code-quote">'index'</span>);
    }
...</pre>
</div></div>
<p>It will create cache for each group permutation   <span class="image-wrap" style=""><img src="http://forge.prestashop.com/images/icons/emoticons/smile.png" height="16" width="16" style="border: 0px solid black" /></span></p>

**Comment by prestonBot** [comment]
> This comment has been migrated from the Forge. Read the [original comment here](http://forge.prestashop.com/browse/BOOM-3633#comment-171558).

- _**Author:**_ marion_francois
- _**Created at:**_ Fri, 13 Jul 2018 10:38:43 +0200

<p>Thanks for diving into this <a href="http://forge.prestashop.com/secure/ViewProfile.jspa?name=iKaemos" class="user-hover" rel="iKaemos">Cyril DELERY</a></p>

<p>Would you be willing to make a pull request on GitHub with your code suggestion?<br/>
 <a href="https://github.com/PrestaShop/PrestaShop/tree/develop" class="external-link" rel="nofollow">https://github.com/PrestaShop/PrestaShop/tree/develop</a></p>

<p>Thank you!</p>

**Comment by prestonBot** [comment]
> This comment has been migrated from the Forge. Read the [original comment here](http://forge.prestashop.com/browse/BOOM-3633#comment-171574).

- _**Author:**_ ikaemos
- _**Created at:**_ Fri, 13 Jul 2018 12:00:44 +0200

<p>Hi Marion!</p>

<p>I will, once my current dev is done <img class="emoticon" src="http://forge.prestashop.com/images/icons/emoticons/wink.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p> </p>

<p>Thank you!</p>

**Comment by prestonBot** [comment]
> This comment has been migrated from the Forge. Read the [original comment here](http://forge.prestashop.com/browse/BOOM-3633#comment-171601).

- _**Author:**_ dubaku
- _**Created at:**_ Fri, 13 Jul 2018 15:52:28 +0200

<p>Wow!</p>

<p>First of all THANKS, I'm waited for this fix 10000000years ago <img class="emoticon" src="http://forge.prestashop.com/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>So I made the changes for modules/ps_mainmenu/ps_mainmenu.php</p>

<p>After I checked the categories and categories are works now, but if I'm logged in and I'm in a category which is prevented from visitors and I'm doing a logout I don't get the usual error message(You don't have the permission to acces this page), instead I get this error:</p>
<ol>
	<li>in b2019fdfedd1ef8dd9605f66e2a0617426c31623.file.category.tpl.php line 232</li>
	<li>at ErrorHandler->handleError('8', 'Undefined index: category', '*****/app/cache/dev/smarty/compile/layouts_layout_left_column_tpl/b2/01/9f/b2019fdfedd1ef8dd9605f66e2a0617426c31623.file.category.tpl.php', '232', <em>array</em>('<em>smarty_tpl' => _object</em>(SmartyDevTemplate), '<em>tpl_stack' => _array</em>())) in b2019fdfedd1ef8dd9605f66e2a0617426c31623.file.category.tpl.php line 232</li>
	<li>at content_5b48ab355ead47_01510754(<em>object</em>(SmartyDevTemplate)) in smarty_internal_templatebase.php line 188</li>
	<li>at Smarty_Internal_TemplateBase->fetch('catalog/listing/category.tpl', <em>null</em>, 'layouts/layout-left-column.tpl', <em>null</em>, <em>false</em>, <em>true</em>, <em>false</em>) in SmartyDev.php line 41</li>
	<li>at SmartyDev->fetch('catalog/listing/category.tpl', <em>null</em>, 'layouts/layout-left-column.tpl') in FrontController.php line 684</li>
	<li>at FrontControllerCore->smartyOutputContent('catalog/listing/category.tpl') in FrontController.php line 668</li>
	<li>at FrontControllerCore->display() in Controller.php line 273</li>
	<li>at ControllerCore->run() in Dispatcher.php line 428</li>
	<li>at DispatcherCore->dispatch() in index.php line 28</li>
</ol>

**Comment by prestonBot** [comment]
> This comment has been migrated from the Forge. Read the [original comment here](http://forge.prestashop.com/browse/BOOM-3633#comment-171628).

- _**Author:**_ ikaemos
- _**Created at:**_ Fri, 13 Jul 2018 18:51:27 +0200

<p>Hi <img class="emoticon" src="http://forge.prestashop.com/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>For this error you need to modify the two following methods in CategoryController :</p>

<p>If you want to override it you have to create the full path before : "/override/controllers/front/listing/" (it doesn't exist, so you have to create it to prevent a crash)</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
<span class="code-keyword">public</span> function initContent()
{
    parent::initContent();
    <span class="code-keyword">if</span> ($<span class="code-keyword">this</span>->category->checkAccess($<span class="code-keyword">this</span>->context->customer->id))
$<span class="code-keyword">this</span>->doProductSearch(<span class="code-quote">'catalog/listing/category'</span>, array(<span class="code-quote">'entity'</span> => <span class="code-quote">'category'</span>, <span class="code-quote">'id'</span> => $<span class="code-keyword">this</span>->category->id));
    <span class="code-keyword">else</span>
    {
header(<span class="code-quote">'HTTP/1.1 403 Forbidden'</span>);
header(<span class="code-quote">'Status: 403 Forbidden'</span>);
    }
}

<span class="code-keyword">public</span> function getLayout()
{
    <span class="code-keyword">if</span> (!($<span class="code-keyword">this</span>->category->checkAccess($<span class="code-keyword">this</span>->context->customer->id)))
<span class="code-keyword">return</span> <span class="code-quote">'layouts/layout-full-width.tpl'</span>;

    <span class="code-keyword">return</span> parent::getLayout();
}
</pre>
</div></div>

**Comment by prestonBot** [comment]
> This comment has been migrated from the Forge. Read the [original comment here](http://forge.prestashop.com/browse/BOOM-3633#comment-171632).

- _**Author:**_ dubaku
- _**Created at:**_ Fri, 13 Jul 2018 19:23:15 +0200

<p>Thanks Cyril DELERY!</p>

<p>Nice one! The error is gone!</p>

<p>Have a nice weekend!</p>

**Comment by prestonBot** [comment]
> This comment has been migrated from the Forge. Read the [original comment here](http://forge.prestashop.com/browse/BOOM-3633#comment-171633).

- _**Author:**_ ikaemos
- _**Created at:**_ Fri, 13 Jul 2018 20:16:32 +0200

<p><img class="emoticon" src="http://forge.prestashop.com/images/icons/emoticons/wink.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by prestonBot** [comment]
> This comment has been migrated from the Forge. Read the [original comment here](http://forge.prestashop.com/browse/BOOM-3633#comment-172023).

- _**Author:**_ dubaku
- _**Created at:**_ Wed, 18 Jul 2018 14:25:48 +0200

<p>Now I get this errormessage when I create a new product/save existing product or preview:</p>

<p>ContextErrorException in ps_mainmenu.php line 985: Catchable Fatal Error: Argument 1 passed to Ps_MainMenu::cleanMenuCacheDirectory() must be an instance of string, string given, called in ***/modules/ps_mainmenu/ps_mainmenu.php on line 982 and defined</p>

<p>in ps_mainmenu.php line 985<br/>
 at ErrorHandler->handleError('4096', 'Argument 1 passed to Ps_MainMenu::cleanMenuCacheDirectory() must be an instance of string, string given, called in /***<b>/modules/ps_mainmenu/ps_mainmenu.php on line 982 and defined', '/</b>***/modules/ps_mainmenu/ps_mainmenu.php', '985', array()) in ps_mainmenu.php line 985<br/>
 at Ps_MainMenu->cleanMenuCacheDirectory('/****/app/cache/dev/ps_mainmenu') in ps_mainmenu.php line 982<br/>
 at Ps_MainMenu->clearMenuCache() in ps_mainmenu.php line 939<br/>
 at Ps_MainMenu->hookActionObjectProductAddAfter(array('object' => object(Product), 'cookie' => object(Cookie), 'cart' => null, 'altern' => '1')) in Hook.php line 924<br/>
 at HookCore::coreCallHook(object(Ps_MainMenu), 'hookactionObjectProductAddAfter', array('object' => object(Product), 'cookie' => object(Cookie), 'cart' => null, 'altern' => '1')) in Hook.php line 328<br/>
 at HookCore::callHookOn(object(Ps_MainMenu), 'actionObjectProductAddAfter', array('object' => object(Product), 'cookie' => object(Cookie), 'cart' => null, 'altern' => '1')) in Hook.php line 860<br/>
 at HookCore::exec('actionObjectProductAddAfter', array('object' => object(Product))) in ObjectModel.php line 579<br/>
 at ObjectModelCore->add(true, false) in Product.php line 596<br/>
 at ProductCore->add(true, false) in ObjectModel.php line 487<br/>
 at ObjectModelCore->save() in ProductController.php line 372<br/>
 at ProductController->newAction()<br/>
 at call_user_func_array(array(object(ProductController), 'newAction'), array()) in bootstrap.php.cache line 3253<br/>
 at HttpKernel->handleRaw(object(Request), '1') in bootstrap.php.cache line 3212<br/>
 at HttpKernel->handle(object(Request), '1', false) in bootstrap.php.cache line 3366<br/>
 at ContainerAwareHttpKernel->handle(object(Request), '1', false) in bootstrap.php.cache line 2566<br/>
 at Kernel->handle(object(Request), '1', false) in index.php line 86</p>

**Comment by prestonBot** [comment]
> This comment has been migrated from the Forge. Read the [original comment here](http://forge.prestashop.com/browse/BOOM-3633#comment-172105).

- _**Author:**_ ikaemos
- _**Created at:**_ Thu, 19 Jul 2018 00:39:02 +0200

<p>Hi <img class="emoticon" src="http://forge.prestashop.com/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p> </p>

<p>It may be related to your php version. Anyway it's a weird crash:</p>

<p>"must be an instance of string, string given" <img class="emoticon" src="http://forge.prestashop.com/images/icons/emoticons/biggrin.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p> </p>

<p>You can try this:</p>

<p>Replace <br/>
private function cleanMenuCacheDirectory(string $dir)<br/>
to<br/>
private function cleanMenuCacheDirectory($dir)</p>

**Comment by prestonBot** [comment]
> This comment has been migrated from the Forge. Read the [original comment here](http://forge.prestashop.com/browse/BOOM-3633#comment-172189).

- _**Author:**_ dubaku
- _**Created at:**_ Thu, 19 Jul 2018 14:56:13 +0200

<p>Thanks Cyril DELERY the quick answer/fix, have a nice day!</p>

**Comment by wileeVB** [comment]
Hello, I'm using ver 1.7.4.2.
I need to hide category on the menu until users logged in.
I made changes to ps_mainmenu and CategoryController as above instructed.
Its good before logging on, but after login it gave me a "Array to string conversion" error.

Please see attached screen shot.
![errormsg](https://user-images.githubusercontent.com/46695504/51408377-5a033900-1b13-11e9-8680-0160bb9b2403.PNG)

Please help.
Thanks.

**Comment by khouloudbelguith** [comment]
Hi @wileeVB,

Please be aware that the forge should be used to report issues only. If you need help, you can ask on the Forum (https://www.prestashop.com/forums/), Gitter (https://gitter.im/PrestaShop/General) or Stack Overflow (https://stackoverflow.com/questions/tagged/prestashop).
Thanks for your understanding!

**Comment by apcloic** [comment]
Hi,

It looks like the fix works well for product category items but not for other kind of items.
For example, I would like to hide a link to a quick order module page which should be only visible for some customer groups.
For now, even if the module is disabled for visitors and clients and nothing happens (as expected) when they click on it via menu link, I think it should not be visible at all for these groups.

**Comment by lpoczatek** [comment]
Hi guys,
I've made fix for me in another way. I overrided one file.

```php
<?php

class Ps_MainMenuOverride extends Ps_MainMenu
{
    public function getWidgetVariables($hookName, array $configuration)
    {
        $id_lang = $this->context->language->id;
        $id_shop = $this->context->shop->id;
        // fix for: https://github.com/PrestaShop/PrestaShop/issues/9924
        $customerGroups = Customer::getGroupsStatic((int)$this->context->customer->id);

        $key = self::MENU_JSON_CACHE_KEY . '_' . $id_lang . '_' . $id_shop . '_' . md5(implode('', $customerGroups)) . '.json';
        $cacheDir = $this->getCacheDirectory();
        $cacheFile = $cacheDir . DIRECTORY_SEPARATOR . $key;
        $menu = json_decode(@file_get_contents($cacheFile), true);
        if (!is_array($menu) || json_last_error() !== JSON_ERROR_NONE) {
            $menu = $this->makeMenu();
            if (!is_dir($cacheDir)) {
                mkdir($cacheDir);
            }
            file_put_contents($cacheFile, json_encode($menu));
        }

        $page_identifier = $this->getCurrentPageIdentifier();
        // Mark the current page
        return $this->mapTree(function (array $node) use ($page_identifier) {
            $node['current'] = ($page_identifier === $node['page_identifier']);
            return $node;
        }, $menu);
    }

    protected function makeMenu()
    {
        $root_node = $this->makeNode([
            'label' => null,
            'type'  => 'root',
            'children' => []
        ]);

        $menu_items = $this->getMenuItems();
        $id_lang = (int)$this->context->language->id;
        $id_shop = (int)Shop::getContextShopID();

        foreach ($menu_items as $item) {
            if (!$item) {
                continue;
            }

            preg_match($this->pattern, $item, $value);
            $id = (int)substr($item, strlen($value[1]), strlen($item));

            switch (substr($item, 0, strlen($value[1]))) {
                case 'CAT':
                    // fix for: https://github.com/PrestaShop/PrestaShop/issues/9924
                    $customerGroups = Customer::getGroupsStatic((int)$this->context->customer->id);
                    $sqlFilter = ' AND c.id_category IN (
                        SELECT id_category
                        FROM `'._DB_PREFIX_.'category_group`
                        WHERE `id_group` IN ('.pSQL(implode(', ', $customerGroups)).')
                    ) ';

                    $categories = $this->generateCategoriesMenu(
                        Category::getNestedCategories($id, $id_lang, false, $this->user_groups, true, $sqlFilter)
                    );
                    $root_node['children'] = array_merge($root_node['children'], $categories);
                    break;

                case 'PRD':
                    $product = new Product((int)$id, true, (int)$id_lang);
                    if ($product->id) {
                        $root_node['children'][] = $this->makeNode([
                            'type' => 'product',
                            'page_identifier' => 'product-' . $product->id,
                            'label' => $product->name,
                            'url' => $product->getLink(),
                        ]);
                    }
                    break;

                case 'CMS':
                    $cms = CMS::getLinks((int)$id_lang, array($id));
                    if (count($cms)) {
                        $root_node['children'][] = $this->makeNode([
                            'type' => 'cms-page',
                            'page_identifier' => 'cms-page-' . $id,
                            'label' => $cms[0]['meta_title'],
                            'url' => $cms[0]['link']
                        ]);
                    }
                    break;

                case 'CMS_CAT':
                    $root_node['children'][] = $this->generateCMSCategoriesMenu((int)$id, (int)$id_lang);
                    break;

                // Case to handle the option to show all Manufacturers
                case 'ALLMAN':
                    $children = array_map(function ($manufacturer) use ($id_lang) {
                        return $this->makeNode([
                            'type' => 'manufacturer',
                            'page_identifier' => 'manufacturer-' . $manufacturer['id_manufacturer'],
                            'label' => $manufacturer['name'],
                            'url' => $this->context->link->getManufacturerLink(
                                new Manufacturer($manufacturer['id_manufacturer'], $id_lang),
                                null,
                                $id_lang
                            )
                        ]);
                    }, Manufacturer::getManufacturers());

                    $root_node['children'][] = $this->makeNode([
                        'type' => 'manufacturers',
                        'page_identifier' => 'manufacturers',
                        'label' => $this->trans('All brands', array(), 'Modules.Mainmenu.Admin'),
                        'url' => $this->context->link->getPageLink('manufacturer'),
                        'children' => $children
                    ]);
                    break;

                case 'MAN':
                    $manufacturer = new Manufacturer($id, $id_lang);
                    if ($manufacturer->id) {
                        $root_node['children'][] = $this->makeNode([
                            'type' => 'manufacturer',
                            'page_identifier' => 'manufacturer-' . $manufacturer->id,
                            'label' => $manufacturer->name,
                            'url' => $this->context->link->getManufacturerLink(
                                $manufacturer,
                                null,
                                $id_lang
                            )
                        ]);
                    }
                    break;

                // Case to handle the option to show all Suppliers
                case 'ALLSUP':
                    $children = array_map(function ($supplier) use ($id_lang) {
                        return $this->makeNode([
                            'type' => 'supplier',
                            'page_identifier' => 'supplier-' . $supplier['id_supplier'],
                            'label' => $supplier['name'],
                            'url' => $this->context->link->getSupplierLink(
                                new Supplier($supplier['id_supplier'], $id_lang),
                                null,
                                $id_lang
                            )
                        ]);
                    }, Supplier::getSuppliers());

                    $root_node['children'][] = $this->makeNode([
                        'type' => 'suppliers',
                        'page_identifier' => 'suppliers',
                        'label' => $this->trans('All suppliers', array(), 'Modules.Mainmenu.Admin'),
                        'url' => $this->context->link->getPageLink('supplier'),
                        'children' => $children
                    ]);
                    break;

                case 'SUP':
                    $supplier = new Supplier($id, $id_lang);
                    if ($supplier->id) {
                        $root_node['children'][] = $this->makeNode([
                            'type' => 'supplier',
                            'page_identifier' => 'supplier-' . $supplier->id,
                            'label' => $supplier->name,
                            'url' => $this->context->link->getSupplierLink(
                                $supplier,
                                null,
                                $id_lang
                            )
                        ]);
                    }
                    break;

                case 'SHOP':
                    $shop = new Shop((int)$id);
                    if (Validate::isLoadedObject($shop)) {
                        $root_node['children'][] = $this->makeNode([
                            'type' => 'shop',
                            'page_identifier' => 'shop-' . $id,
                            'label' => $shop->name,
                            'url' => $shop->getBaseURL(),
                        ]);
                    }
                    break;
                case 'LNK':
                    $link = Ps_MenuTopLinks::get($id, $id_lang, $id_shop);
                    if (!empty($link)) {
                        if (!isset($link[0]['label']) || ($link[0]['label'] == '')) {
                            $default_language = Configuration::get('PS_LANG_DEFAULT');
                            $link = Ps_MenuTopLinks::get($link[0]['id_linksmenutop'], $default_language, (int)Shop::getContextShopID());
                        }
                        $root_node['children'][] = $this->makeNode([
                            'type' => 'link',
                            'page_identifier' => 'lnk-' . Tools::str2url($link[0]['label']),
                            'label' => $link[0]['label'],
                            'url' => $link[0]['link'],
                            'open_in_new_window' => $link[0]['new_window']
                        ]);
                    }
                    break;
            }
        }

        return $this->mapTree(function ($node, $depth) {
            $node['depth'] = $depth;
            return $node;
        }, $root_node);
    }

    private function getCurrentPageIdentifier()
    {
        $controllerName = Dispatcher::getInstance()->getController();
        if ($controllerName === 'cms' && ($id = Tools::getValue('id_cms'))) {
            return 'cms-page-' . $id;
        } else if ($controllerName === 'category' && ($id = Tools::getValue('id_category'))) {
            return 'category-' . $id;
        } else if ($controllerName === 'cms' && ($id = Tools::getValue('id_cms_category'))) {
            return 'cms-category-' . $id;
        } else if ($controllerName === 'manufacturer' && ($id = Tools::getValue('id_manufacturer'))) {
            return 'manufacturer-' . $id;
        } else if ($controllerName === 'manufacturer') {
            return 'manufacturers';
        } else if ($controllerName === 'supplier' && ($id = Tools::getValue('id_supplier'))) {
            return 'supplier-' . $id;
        } else if ($controllerName === 'supplier') {
            return 'suppliers';
        } else if ($controllerName === 'product' && ($id = Tools::getValue('id_product'))) {
            return 'product-' . $id;
        } else if ($controllerName === 'index') {
            return 'shop-' . $this->context->shop->id;
        } else {
            $scheme = 'http';
            if (array_key_exists('REQUEST_SCHEME', $_SERVER)) {
                $scheme = $_SERVER['REQUEST_SCHEME'];
            }
            return "$scheme://$_SERVER[HTTP_HOST]$_SERVER[REQUEST_URI]";
        }
    }
}
```

**Comment by khouloudbelguith** [comment]
Hi @lpoczatek,


Would you be willing to make a pull request on GitHub with a code suggestion?
https://github.com/PrestaShop/PrestaShop/tree/develop
https://github.com/PrestaShop/ps_mainmenu
Thank you!

**Comment by lpoczatek** [comment]
Hi,
I edited my previous post with current version of code. Just one file only changed/added :).
I don't have time to make pull request (I don't have configured repo, etc.), please feel free to use my suggestion if you want :). I marked my changes by comments like this:
`// fix for: https://github.com/PrestaShop/PrestaShop/issues/9924`
So in pull request will be probably changed only 6-7 lines I think but now I have more code copied because of override.

**Comment by rvalery** [comment]
Hello,

I ran into the same problem that is described in this topic. I saw the comments and the 2 different solutions that were proposed. None of them seem to have been merged into the master branch of the module, although I have seen some commits for this on the dev branch.

I think I have found a solution for this problem that is simpler than the proposed modifications. I simply initialize the "user_groups" property that was already defined in the Ps_MainMenu class, but never initialized. This property was already used in the function makeMenu to restrict category list to only those accessible for the user. So I only initialize the property, and use it to create the cache key for the menu. This is done in the function getWidgetVariables :

```
    public function getWidgetVariables($hookName, array $configuration)
    {
        $id_lang = $this->context->language->id;
        $id_shop = $this->context->shop->id;
        
        $this->user_groups = Customer::getGroupsStatic($this->context->customer->id);
        $groupsKey = empty($this->user_groups) ? '' : '_' . join("_", $this->user_groups);
        $key = self::MENU_JSON_CACHE_KEY . '_' . $id_lang . '_' . $id_shop . $groupsKey . '.json';
        $cacheDir = $this->getCacheDirectory();
        $cacheFile = $cacheDir . DIRECTORY_SEPARATOR . $key;
        $menu = json_decode(@file_get_contents($cacheFile), true);
        if (!is_array($menu) || json_last_error() !== JSON_ERROR_NONE) {
            $menu = $this->makeMenu();
            if (!is_dir($cacheDir)) {
                mkdir($cacheDir);
            }
            file_put_contents($cacheFile, json_encode($menu));
        }

        $page_identifier = $this->getCurrentPageIdentifier();
        // Mark the current page
        return $this->mapTree(function (array $node) use ($page_identifier) {
            $node['current'] = ($page_identifier === $node['page_identifier']);
            return $node;
        }, $menu);
    }
```

I have created a pull request for this.

Until it is fixed in the module, I have created an override file (override/modules/ps_mainmenu/ps_mainmenu.php) with the following content :
```
<?php

class Ps_MainMenuOverride extends Ps_MainMenu
{

    private function getCurrentPageIdentifier()
    {
        $controllerName = Dispatcher::getInstance()->getController();
        if ($controllerName === 'cms' && ($id = Tools::getValue('id_cms'))) {
            return 'cms-page-' . $id;
        } else if ($controllerName === 'category' && ($id = Tools::getValue('id_category'))) {
            return 'category-' . $id;
        } else if ($controllerName === 'cms' && ($id = Tools::getValue('id_cms_category'))) {
            return 'cms-category-' . $id;
        } else if ($controllerName === 'manufacturer' && ($id = Tools::getValue('id_manufacturer'))) {
            return 'manufacturer-' . $id;
        } else if ($controllerName === 'manufacturer') {
            return 'manufacturers';
        } else if ($controllerName === 'supplier' && ($id = Tools::getValue('id_supplier'))) {
            return 'supplier-' . $id;
        } else if ($controllerName === 'supplier') {
            return 'suppliers';
        } else if ($controllerName === 'product' && ($id = Tools::getValue('id_product'))) {
            return 'product-' . $id;
        } else if ($controllerName === 'index') {
            return 'shop-' . $this->context->shop->id;
        } else {
            $scheme = 'http';
            if (array_key_exists('REQUEST_SCHEME', $_SERVER)) {
                $scheme = $_SERVER['REQUEST_SCHEME'];
            }
            return "$scheme://$_SERVER[HTTP_HOST]$_SERVER[REQUEST_URI]";
        }
    }
    
    public function getWidgetVariables($hookName, array $configuration)
    {
        $id_lang = $this->context->language->id;
        $id_shop = $this->context->shop->id;
        
        $this->user_groups = Customer::getGroupsStatic($this->context->customer->id);
        $groupsKey = empty($this->user_groups) ? '' : '_' . join("_", $this->user_groups);
        $key = self::MENU_JSON_CACHE_KEY . '_' . $id_lang . '_' . $id_shop . $groupsKey . '.json';
        $cacheDir = $this->getCacheDirectory();
        $cacheFile = $cacheDir . DIRECTORY_SEPARATOR . $key;
        $menu = json_decode(@file_get_contents($cacheFile), true);
        if (!is_array($menu) || json_last_error() !== JSON_ERROR_NONE) {
            $menu = $this->makeMenu();
            if (!is_dir($cacheDir)) {
                mkdir($cacheDir);
            }
            file_put_contents($cacheFile, json_encode($menu));
        }

        $page_identifier = $this->getCurrentPageIdentifier();
        // Mark the current page
        return $this->mapTree(function (array $node) use ($page_identifier) {
            $node['current'] = ($page_identifier === $node['page_identifier']);
            return $node;
        }, $menu);
    }
}
```

Note that the function getCurrentPageIdentifier() has been copied directly from original class, because this method is private. It does not need any change though, the only changes required are the three lines using "user_groups" in getWidgetVariables.
